zcash_primitives/src/transaction/components/sapling/builder.rs

```diff
@@ -373,204 +462,401 @@ impl<P: consensus::Parameters> SaplingBuilder<P> {
+            |auth: InProgress<P, Unsigned>| InProgress {
+                sigs: PartiallyAuthorized {
+                    binding_signature: auth.sigs.bsk.sign(
+                        &sighash,
+                        &mut rng,
+                        VALUE_COMMITMENT_RANDOMNESS_GENERATOR,
+                    ),
+                    sighash,
+                },
+                _proof_state: PhantomData::default(),
+            },
+        ))
+    }
+}
+
+impl Bundle<InProgress<Proven, Unsigned>> {
+    /// Applies signatures to this bundle, in order to authorize it.
+    ///
+    /// This is a helper method that wraps [`Bundle::prepare`], [`Bundle::sign`], and
+    /// [`Bundle::finalize`].
+    pub fn apply_signatures<R: RngCore + CryptoRng>(
+        self,
+        mut rng: R,
+        sighash: [u8; 32],
+        signing_keys: &[PrivateKey],
+    ) -> Result<Bundle<Authorized>, Error> {
+        signing_keys
+            .iter()
+            .fold(self.prepare(&mut rng, sighash), |partial, ask| {
+                partial.sign(&mut rng, ask)
+            })
+            .finalize()
+    }
+}
+
+impl<P: InProgressProofs> Bundle<InProgress<P, PartiallyAuthorized>> {
+    /// Signs this bundle with the given [`SpendAuthorizingKey`].
+    ///
+    /// This will apply signatures for all notes controlled by this spending key.
+    pub fn sign<R: RngCore + CryptoRng>(self, mut rng: R, ask: &PrivateKey) -> Self {
+        let expected_ak = PublicKey::from_private(ask, SPENDING_KEY_GENERATOR);
+        let sighash = self.authorization.sigs.sighash;
+        self.map_authorization((
+            |proof| proof,
+            |proof| proof,
+            |maybe| match maybe {
+                MaybeSigned::SigningMetadata(parts) if parts.ak.0 == expected_ak.0 => {
+                    MaybeSigned::Signature(spend_sig_internal(ask, parts.alpha, &sighash, &mut rng))
+                }
+                s => s,
+            },
+            |partial| partial,
+        ))
+    }
+    /// Appends externally computed [`Signature`]s.
+    ///
+    /// Each signature will be applied to the one input for which it is valid. An error
+    /// will be returned if the signature is not valid for any inputs, or if it is valid
+    /// for more than one input.
+    ///
+    /// [`Signature`]: redpallas::Signature
+    pub fn append_signatures(self, signatures: &[Signature]) -> Result<Self, Error> {
+        signatures.iter().try_fold(self, Self::append_signature)
+    }
+
+    fn append_signature(self, signature: &Signature) -> Result<Self, Error> {
+        let sighash = self.authorization.sigs.sighash;
+        let mut signature_valid_for = 0usize;
+        let bundle = self.map_authorization((
+            |proof| proof,
+            |proof| proof,
+            |maybe| match maybe {
+                MaybeSigned::SigningMetadata(parts) => {
+                    let rk = parts.ak.randomize(parts.alpha, SPENDING_KEY_GENERATOR);
+                    // TODO: Fix msg
```

Comment:
I don't understand this TODO; can it be fixed now, or if not can what it's trying to say be clarified?

Reply:
This was a TODO I forgot I hadn't done (because COVID). I've implemented a temporary fix; the real fix is to finally migrate to the `redjubjub` crate.